### PR TITLE
Enhance piecewise to allow functions to have arbitrary number of input arguments

### PR DESCRIPTION
**Pull request by pbrod**
Made piecewise function simpler and more general. The new piecewise function allows functions of the type f(x0, x1,.. ,xn) i.e. to have arbitrary number of input arguments that are evaluated conditionally.

The old piecewise function did allow condlist to be a single bool list/array or a single int array 
(That feature was not documented in the docstring of the function, but there were tests testing this functionality.) Question is if this feature should be deprecated. For now this feature is kept in the new version. 

See also: http://www.mail-archive.com/numpy-discussion@scipy.org/msg46304.html
### Examples that work both in previous and current version

```
>>> x = np.linspace(-2,2,5)
>>> np.piecewise(x, [x<1], [1]) 
array([ 1.,  1.,  1.,  0.,  0.])

>>> np.piecewise((x,), [x < 0, x >= 0], [lambda x: -x, lambda x: x])
array([ 2.,  1.,  0.,  1.,  2.])
```

can be written as

```
>>> np.piecewise(x,[x < 0, x >= 0], [lambda x: -x, lambda x: x])
array([ 2.,  1.,  0.,  1.,  2.])
```

or

```
>>> np.piecewise(x, [x < 0,], [lambda x: -x, lambda x: x])
array([ 2.,  1.,  0.,  1.,  2.])
```
### Example that still works but should possibly be deprecated:

```
>>> np.piecewise(x, x < 0, [lambda x: -x, lambda x: x])
array([ 2.,  1.,  0.,  1.,  2.])
```
### Examples demonstrating the new functionality

```
>>> X,Y = np.meshgrid(x,x)
>>> np.piecewise((X,Y), [X*Y<0,], [lambda x,y: -x*y, lambda x,y: x*y])
array([[ 4.,  2., -0.,  2.,  4.],
       [ 2.,  1., -0.,  1.,  2.],
       [-0., -0.,  0.,  0.,  0.],
       [ 2.,  1.,  0.,  1.,  2.],
       [ 4.,  2.,  0.,  2.,  4.]])

>>> np.piecewise((X,Y),[X*Y<-0.5, X*Y>0.5], [lambda x,y: -x*y, lambda x,y: x*y, np.nan])
array([[  4.,   2.,  nan,   2.,   4.],
       [  2.,   1.,  nan,   1.,   2.],
       [ nan,  nan,  nan,  nan,  nan],
       [  2.,   1.,  nan,   1.,   2.],
       [  4.,   2.,  nan,   2.,   4.]])
```

### CODE REVIEW

**Comment by charris** [comment]
We can't break backward compatibility here. Either a new function is needed, or the enhanced function needs to be fixed to work with old arguments.

**Comment by pbrod** [comment]
According to the docstring of the old function:
 condlist : list of bool arrays
        Each boolean array corresponds to a function in `funclist`.  Wherever
        `condlist[i]` is True, `funclist[i](x)` is used as the output value.

Thus the example

    x = np.linspace(-2,2,5)
    np.piecewise(x, x < 0, [lambda x: -x, lambda x: x])

violates the condition that condlist should be a list. Here condlist is a ndarray. The old function allowed this calling syntax, but it is clearly an undocumented feature of the old function. The question is if that is a wanted behavior?

**Comment by pbrod** [comment]
This version of 'piecewice' function resolves the problems mentioned in ticket #5729 and #5737.
@charris I need advice on how to proceed with this.

**Comment by hunterakins** [comment]
Seems this didn't merge...would be a useful enhancement. Any update on this?

**Comment by hunterakins** [comment]
That's alright.
I still don't understand how that has to do with np.full or defining
conditional functions that work on meshgrid 2D arrays.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -838,7 +838,19 @@ class ndarray is returned.
     return a
 
 
-def piecewise(x, condlist, funclist, *args, **kw):
+def valarray(shape, value=0.0, typecode=None):
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/lib/function_base.py]: I'm commenting on `valarray`, not `piecewise`

**Comment by anirudh2290** [comment]
@pbrod This has been open for a while. Do you plan to come back to this ?

**Comment by pbrod** [comment]
@anirudh2290 Well, I am happy to complete this, but I need guidance on how to proceed.

**Comment by anirudh2290** [comment]
@pbrod According to me, looks like this enhancement would be a good to have. 
As for the further direction, as @eric-wieser suggests we should first deprecate the tuple input which is just converted to a single ndarray and add a FutureWarning for the same. We can add this new behavior after 2 releases. 

> Thus the example

> x = np.linspace(-2,2,5)
> np.piecewise(x, x < 0, [lambda x: -x, lambda x: x])
> violates the condition that condlist should be a list. Here condlist is a ndarray. The old function 
> allowed this calling syntax, but it is clearly an undocumented feature of the old function. The > question is if that is a wanted behavior?

Since single condition is promoted to a list of one condition, I won't mind keeping it as it is and documenting this behavior.

cc @eric-wieser for a second look.

**Comment by eric-wieser** [comment]
Can we come up with a new spelling that doesn't overlap the old one?

As a reminder, we currently have:
```python
np.piecewise(
    arr_like,
    [cond1, cond2, cond3],
    [func1, func2, func3, func_default],
    *func_args, **func_kwargs)
```
One option might be pulling some magic with `__getitem__`:
```python
np.piecewise[2](
    arr_like1,
    arr_like2,
    [cond1, cond2, cond3],
    [func1, func2, func3, func_default],
    *func_args, **func_kwargs)
```
or by currying (noting that 2-argument invocations are currently not allowed):
```python
np.piecewise(
    [cond1, cond2, cond3],
    [func1, func2, func3, func_default])(arr_like1, arr_like2)
```
Will edit this post if I think of more...

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -847,34 +859,33 @@ def piecewise(x, condlist, funclist, *args, **kw):
...
+        `condlist[i]` is True, `funclist[i](x0,x1,...,xn)` is used as the
+        output value. Each boolean array in `condlist` selects a piece of `xi`,
+        and should therefore be of the same shape as `xi`.
 
         The length of `condlist` must correspond to that of `funclist`.
         If one extra function is given, i.e. if
         ``len(funclist) - len(condlist) == 1``, then that extra function
         is the default value, used wherever all conditions are false.
-    funclist : list of callables, f(x,*args,**kw), or scalars
+    funclist : list of callables, f(*(xi + args), **kw), or scalars
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:874]:
```suggestion
    funclist : list of callables, f(*xi, *args, **kw), or scalars
```
is also legal and probably clearer

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -912,49 +923,84 @@ def piecewise(x, condlist, funclist, *args, **kw):
...
+    dtype = np.result_type(fill_value, *xi)
+
+    if xi[0].ndim == 0:
+        # This may be common, so avoid the call.
+        result_shape = condlist[0].shape
+    else:
+        result_shape = np.broadcast_arrays(xi[0], condlist[0])[0].shape
+    # Use np.copyto to burn each funclist array onto result, using the
+    # corresponding condlist as a boolean mask. This is done in reverse
+    # order since the first choice should take precedence.
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:995]: xref #16475, this is a behavior change too

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -912,49 +923,84 @@ def piecewise(x, condlist, funclist, *args, **kw):
...
+                deprecated_ints = True
+            else:
+                raise ValueError(
+                    'invalid entry in condlist: should be boolean ndarray')
+
+    if deprecated_ints:
+        msg = "piecewise condlists containing integer ndarrays is deprecated " \
+            "and will be removed in the future. Use `.astype(bool)` to " \
+            "convert to bools."
+        warnings.warn(msg, DeprecationWarning)
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:982]: This change is probably sensible to extract to its own PR. Note this is missing `stacklevel=2`.

File: numpy/lib/tests/test_function_base.py [hunk 5]
```diff
@@ -1846,27 +1853,123 @@ def test_two_conditions(self):
...
         x = np.array(3)
-        y = piecewise(x, x > 3, [4, 0])
+        y = piecewise(x, [x > 3], [4, 0])
         assert_(y.ndim == 0)
         assert_(y == 0)
 
         x = 5
         y = piecewise(x, [[True], [False]], [1, 0])
-        assert_(y.ndim == 0)
+        assert_(y.ndim == 1)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:1871]: This is a breaking change as well sadly.

**Review by eric-wieser — COMMENTED** [review]
Looks like there are more behavior changes here than I first spotted

File: numpy/lib/function_base.py (hunk 2 above)
Comment (anirudh2290) [reply numpy/lib/function_base.py:874]: xi may be slightly misleading since these funcs run on indexed array as such which will be 1d and may not be all elements of xi.
